ncm-systemd/src/main/perl/Systemd/UnitFile.pm

```diff
@@ -0,0 +1,444 @@
+                     " (filename $filename) failed: $trd->{fail}.");
+        return;
+    }
+
+    my $changed = $fh->close() ? 1 : 0; # force to 1 or 0
+
+    # if changed, reload daemon
+    if($changed) {
+        # can't do much with return value?
+        CAF::Process->new(
+            [$SYSTEMCTL, $DAEMON_RELOAD],
+            log => $self,
+            )->execute();
+    }
+
+    return $changed;
+}
+
+=pod
+
+=back
+
+=head2 Private methods
+
+=over
+
+=item _prepare_path
+
+Create and return the filename to use,
+and prepare the directory structure if needed.
+
+C<basedir> is the base directory to use, e.g. C<$UNITFILE_DIRECTORY>.
+
+=cut
+
+sub _prepare_path
+{
+    my ($self, $basedir) = @_;
+
+    my $filename;
+
+    my $unitfile = "$basedir/$self->{unit}";
+    my $unitdir = "${unitfile}.d";
+
+    if ($self->{replace}) {
+        # unitdir can't exist
+        return if (! $self->_cleanup($unitdir));
+
+        $filename = $unitfile;
+    } else {
+        # unitfile can't exist
+        return if (! $self->_cleanup($unitfile));
+
+        $filename = "$unitdir/$NOREPLACE_FILENAME";
+        if (! ($self->_directory_exists($unitdir) || $self->_make_directory($unitdir))) {
+            $self->error("Failed to create unitdir $unitdir: $!");
+            return;
+        }
+    };
+
+    return $filename;
+}
+
+=item _hwloc_calc_cpuaffinity
+
+Run C<_hwloc_calc_cpus>, and returns in C<CPUAffinity> format with a reset
+
+=cut
+
+sub _hwloc_calc_cpuaffinity
+{
+    my ($self, $locations) = @_;
+
+    my $cpus = $self->_hwloc_calc_cpus($locations);
+    return if(! defined($cpus));
+
+    # first empty list, to reset all previous defined CPUaffinity settings
+    return [[], $cpus];
+}
+
+
+=item _hwloc_calc_cpus
+
+Run the C<hwloc-calc --physical --intersect PU> command for C<locations>.
+
+Returns arrayref with CPU indices on success, undef otherwise.
+
+=cut
+
+sub _hwloc_calc_cpus
+{
+    my ($self, $locations) = @_;
+
+    local $ENV{HWLOC_HIDE_ERRORS};
+    $ENV{HWLOC_HIDE_ERRORS}=1;
+
+    # pass a copy of the Readonly array, so we can extend it
+    my $proc = CAF::Process->new([@HWLOC_CALC_CPUS], log => $self);
+    $proc->pushargs(@$locations);
+
+    my $output = $proc->output();
+    my @indices = grep { $_ =~ m/^\d+$/ } split(/,/, $output);
+    if (join(',', @indices) eq $output) {
+        return \@indices;
+    } else {
+        $self->error("Unexpected output from $proc: $output");
+        return;
+    }
+}
+
+
+=item _make_variables_custom
+
+A function that return the custom variables hashref to pass as ttoptions.
+(This is a function, not a method).
+
+=cut
+
+sub _make_variables_custom {
+    my $customs = shift;
+    my $ttoptions;
+    $ttoptions->{VARIABLES}->{SYSTEMD}->{CUSTOM} = $customs;
+    return $ttoptions;
+}
+
+# TODO: Move to CAF::AllTheMissingBitsThatLCProvides
+
+# make directory, mkdir -p style, simple mkpath wrapper
+sub _make_directory
```

Comment:
`LC::Check::directory` apparently, https://github.com/quattor/LC/blob/master/src/main/perl/Check.pm#L207 (the pod is very unhelpful to find out it actually creates something)